stdlib/public/core/ClosedRange.swift

```diff
@@ -488,3 +488,30 @@ extension ClosedRange {
 // shorthand. TODO: Add documentation
 public typealias CountableClosedRange<Bound: Strideable> = ClosedRange<Bound>
   where Bound.Stride : SignedInteger
+
+extension ClosedRange: Codable where Bound: Codable {
+  private enum CodingKeys: String, CodingKey {
+    case lowerBound = "from"
+    case upperBound = "through"
+  }
+
+  public init(from decoder: Decoder) throws {
+    let container = try decoder.container(keyedBy: CodingKeys.self)
+    let lowerBound = try container.decode(Bound.self, forKey: .lowerBound)
+    let upperBound = try container.decode(Bound.self, forKey: .upperBound)
+    guard lowerBound <= upperBound else {
+      throw DecodingError.dataCorruptedError(
+        forKey: CodingKeys.upperBound,
+        in: container,
+        debugDescription: "upperBound (through) cannot be less than lowerBound (from)"
+      )
+    }
+    self = lowerBound...upperBound
+  }
+
+  public func encode(to encoder: Encoder) throws {
+    var container = encoder.container(keyedBy: CodingKeys.self)
+    try container.encode(self.lowerBound, forKey: .lowerBound)
+    try container.encode(self.upperBound, forKey: .upperBound)
+  }
```

Comment:
The problem with using unkeyed values is that someone could mix up decoding `Range` with `ClosedRange` which would result in the off by 1 error mentioned in the proposal amendment.

There have been a few discussions on this already which you can find in the [forum thread](https://forums.swift.org/t/range-conform-to-codable/15552/12?u=dlbuckley) and in the [bug report](https://bugs.swift.org/browse/SR-8649). It's probably best to keep the discussion moving there to keep it all in one place.

Reply:
This is after all, the code review, so I would like to keep it focused on the code. I think the overall proposal of adopting `Codable` belongs in the other places (and I support it).

Itai and I had a conversation about this this morning, but I feel very strongly that we should prefer a compact representation for fundamental and mathematically-defined types like this. The ranges with only one value should even use a single value container. 

I don't think that we should add a permanent performance cost to using `Range` in archives to attempt to prevent developers from shooting themselves in the foot. And, as the maintainer of `NSKeyedArchiver` for many years, which is heavily string-key based, the performance cost is real.

Ultimately, `Codable` allows readers of archives to make their own decisions about what type to unarchive as.

Reply:
OK sure, lets keep the code discussion here :)

I totally get where you are coming from in terms of `NSKeyedArchiver`, it makes sense to have as little overhead as possible. I'm willing to change it if needs be.

But with that in mind I'm coming at it from the JSON read/write perspective as thats what I think the original drive behind `Codable` was. Ideally the output needs to not only be efficient, but also obvious to the consumer of the data. JSON, and even PropertyLists typically has quite verbose keys so that someone looking at the data knows what's going on without knowing the implementation detail of what created it.

But lets say that we did use an un-keyed or single value container, this is what it would look like in terms of a struct with a date range property converted into JSON (just stating this for keeping all discussions clear in black and white):

Closed Range: `{"myRange" : ["1970-01-01T02:46:40Z", "1970-01-01T05:33:20Z"]}`
Range: `{"myRange" : ["1970-01-01T02:46:40Z", "1970-01-01T05:33:20Z"]}`
PartialRangeUpTo: `{"myRange" : "1970-01-01T02:46:40Z"}`
PartialRangeThrough: `{"myRange" : "1970-01-01T02:46:40Z"}`
PartialRangeFrom: `{"myRange" : "1970-01-01T02:46:40Z"}`

From the JSON perspective the only way you would be able to distinguish between the range types would be documentation for the payload. Compared to:

Closed Range: `{"myRange" : {"from": "1970-01-01T02:46:40Z", "through": "1970-01-01T05:33:20Z"}}`
Range: `{"myRange" : {"from": "1970-01-01T02:46:40Z", "upTo": "1970-01-01T05:33:20Z"}}`
PartialRangeUpTo: `{"myRange" : {"upTo": "1970-01-01T05:33:20Z"}}`
PartialRangeThrough: `{"myRange" : {"through": "1970-01-01T05:33:20Z"}}`
PartialRangeFrom: `{"myRange" : {"from": "1970-01-01T05:33:20Z"}}`

The values now make much more sense (in my entirely personal point of view).

Also if you look at whats already been done in Swift Foundation the implementations for types such as `Measurement`, `Decimal` etc. all use keyed containers. I accept that they are not mathematically defined types, but I feel that range would be the ugly duckling of the types that conform to `Codable` if it was unkeyed.

The rest of the types in the Standard Library have a natural fit with un-keyed containers, but most in Swift Foundation use keyed containers. The `Range` types are a bit of an odd one out in the Standard Library, I'm not arguing that they shouldn't exist there, more that it hasn't been tackled yet, but thats what this PR is all about I guess :)

Reply:
I think this type is a lot closer in spirit to `AffineTransform`:

```swift
extension AffineTransform : Codable {
    public init(from decoder: Decoder) throws {
        var container = try decoder.unkeyedContainer()
        m11 = try container.decode(CGFloat.self)
        m12 = try container.decode(CGFloat.self)
        m21 = try container.decode(CGFloat.self)
        m22 = try container.decode(CGFloat.self)
        tX  = try container.decode(CGFloat.self)
        tY  = try container.decode(CGFloat.self)
    }
}
```

`CGPoint`:

```swift
extension CGPoint : Codable {
  public init(from decoder: Decoder) throws {
    var container = try decoder.unkeyedContainer()
    let x = try container.decode(CGFloat.self)
    let y = try container.decode(CGFloat.self)
    self.init(x: x, y: y)
  }
```

`CGSize`:

```swift
extension CGSize : Codable {
  public init(from decoder: Decoder) throws {
    var container = try decoder.unkeyedContainer()
    let width = try container.decode(CGFloat.self)
    let height = try container.decode(CGFloat.self)
    self.init(width: width, height: height)
  }
```

(Etc.)

I don't believe that we should impose the cost of carrying documentation strings in the archive itself as a primary goal. String keys can have a documentation as a side benefit. I don't think for mathematically closed types, which can gain no new keys, the documentation side benefit is worth the tradeoff.

When we wrote the `Codable` proposal and implementation, I always had the idea that unkeyed containers should be real peers to the keyed ones, for cases where a compact representation is potentially important.

The reason you find more keyed containers in Foundation is because we have more types there which are not as tightly specified as these mathematical concepts, and we err on the side of keys in that case.

Reply:
I do see your point with this and I do agree. But at the same time I can't help but think of the JSON structure I'm going to be vending to clients with this.

I'm wondering if there is a middle solution here where the default encoding would use unkeyed values but we could somehow set an option for keyed values. It's not really an value encoding strategy as it's not the data thats being encoded, but it's still changing the structure of how the data is represented after encoding.

I'm not sure it's possible given the current API, unless we force some really odd solution where getting a specific type of container would fail forcing you to use one thats specified by a key encoding strategy. Doing something like that would then effect all encoding/decoding methods which is far from desirable as it will break so many things.

Do you have any ideas of any possible solutions for doing something like that?

Reply:
It was a central design point of `Codable` that the type itself gets to decide its encoded format. We only deviate from that for a few key types (the ones with strategies are those types), and it's really more about conforming to a specification that already exists for that data format than attempting to compress the size of the output. That's why the strategies are on specific encoders like `JSONEncoder`.

I'm sure we all have experience with under-documented JSON output, and that is why the idea of putting these strings in the archive is attractive. However, my opinion is that these situations are a failure of specifying the JSON correctly. An application level bug, really. 

We can't really fix that by using string keys for this one type. There will always be more cases where you can't figure out what the intended purpose of a value in JSON is, so this would be at best a partial solution anyway.